### PR TITLE
Add WebTorrent component extension when session restoring webtorrent tabs

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/7330

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: patches/chrome-browser-ui-BUILD.gn.patch [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
-index 4a31f9315ce898ed714a7efca6a8afca9d4baf76..767d1686fa4448a2bac4f1fbc7187c8a37e9e2e6 100644
+index 4a31f9315ce898ed714a7efca6a8afca9d4baf76..8c6423caf1762f5be989eff480a67a9d4757cfb6 100644
 --- a/chrome/browser/ui/BUILD.gn
 +++ b/chrome/browser/ui/BUILD.gn
-@@ -608,6 +608,9 @@ static_library("ui") {
+@@ -2,6 +2,7 @@
+ # Use of this source code is governed by a BSD-style license that can be
+ # found in the LICENSE file.
+ 
++import("//brave/browser/ui/ui.gni")
```
Comment (bridiver) [inline patches/chrome-browser-ui-BUILD.gn.patch:9]: you don't need to import anything for this, just put it in build/config/brave_build.gni

File: browser/ui/ui.gni [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import("//brave/components/brave_webtorrent/browser/buildflags/buildflags.gni")
+
+brave_browser_ui_deps = [
+  "//brave/browser/ui",
+]
+
+if (enable_brave_webtorrent) {
+  brave_browser_ui_deps += [
```
Comment (bridiver) [inline browser/ui/ui.gni:13]: maybe a comment here that it's for the chromium_src files?

File: chromium_src/chrome/browser/ui/browser_tabrestore.cc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#endif
+
+#define AddRestoredTab AddRestoredTab_ChromiumImpl
+#include "../../../../../chrome/browser/ui/browser_tabrestore.cc"
+#undef AddRestoredTab
+
+namespace {
+
+#if BUILDFLAG(ENABLE_BRAVE_WEBTORRENT)
+void MaybeLoadWebtorrent(Browser* browser,
```
Comment (bridiver) [inline chromium_src/chrome/browser/ui/browser_tabrestore.cc:23]: I don't think we want to duplicate the code from BraveWebTorrentNavigationThrottle, maybe you can create a static method on BraveWebTorrentNavigationThrottle or something?

Comment (yrliou) [reply chromium_src/chrome/browser/ui/browser_tabrestore.cc:24]: Addressed in https://github.com/brave/brave-core/pull/6085/commits/fc0d60843960b3a6fbb441219c8997c217babda4

File: patches/chrome-browser-ui-BUILD.gn.patch (hunk 1 above)
Comment (yrliou) [reply patches/chrome-browser-ui-BUILD.gn.patch]: There was an import loop in this case.
As discussed on slack, add a target in `chromium/chrome/browser/ui` with required deps and let `brave/browser/ui` depends on that target in
https://github.com/brave/brave-core/pull/6085/commits/fc0d60843960b3a6fbb441219c8997c217babda4

File: components/brave_webtorrent/browser/webtorrent_util.cc [hunk 4]
```diff
@@ -57,6 +58,17 @@ void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry) {
   registry->RegisterBooleanPref(kWebTorrentEnabled, true);
 }
 
+bool IsWebtorrentPage(const GURL& url) {
```
Comment (bridiver) [inline components/brave_webtorrent/browser/webtorrent_util.cc:61]: nit IsWebtorrentURL

Comment (yrliou) [reply components/brave_webtorrent/browser/webtorrent_util.cc]: Addressed in https://github.com/brave/brave-core/pull/6085/commits/2a601f7357e113e9199e26cad2f33e930083ae5c

**Comment by yrliou** [comment]
CI has unrelated failure:
unit test on android:
```
BatAdsAdConversionsDatabaseTableTest.GetAdConversionsFromCatalogEndpoint
BatAdsCreativeAdNotificationsDatabaseTableTest.GetCreativeAdNotificationsFromCatalogEndpoint
```

dist on Windows:
```
Error information: "Error: SignerSign() failed." (-1073700864/0xc000a000)
01:24:06  SignTool Error: An unexpected internal error has occurred.
```

**Comment by yrliou** [comment]
Latest force push is just for squashing into one commit, will cancel CI builds. Merging.

**Comment by feross** [comment]
Thank you for implementing this @yrliou. You're the best!

**Pull request merged**